src/persistent_merkle_tree/Node.zig

```diff
@@ -612,32 +616,50 @@ pub const Id = enum(u32) {
         errdefer pool.free(path_parents);
 
         var node_id = root_node;
-        // The depth offset to start from, going from the current index to the next
-        // This is the offset of the first shared depth between the two indices
-        var d_offset: Depth = 0; // path_len - depth;
+        // The shared depth between the previous and current index
+        // This is initialized as 0 since the first index has no previous index
+        var d_offset: Depth = 0;
 
-        const states = pool.nodes.items(.state);
-        const lefts = pool.nodes.items(.left);
-        const rights = pool.nodes.items(.right);
+        var states = pool.nodes.items(.state);
+        var lefts = pool.nodes.items(.left);
+        var rights = pool.nodes.items(.right);
 
-        // For each index specified
+        // For each index specified, maintain/update path_lefts and path_rights from root (depth 0) all the way to path_len
+        // but only allocate and update path_parents from the next shared depth to path_len
         for (0..indices.len) |i| {
             // Calculate the gindex bits for the current index
             const index = indices[i];
             const gindex: Gindex = @enumFromInt(@as(Gindex.Uint, @intCast(@intFromEnum(base_gindex) | index)));
 
-            try pool.alloc(path_parents[d_offset..path_len]);
+            // Calculate the depth offset to navigate from current index to the next
+            const next_d_offset = if (i == indices.len - 1)
+                // 0 because there is no next index, it also means node_id is now the new root
+                0
+            else
+                path_len - @as(Depth, @intCast(@bitSizeOf(usize) - @clz(index ^ indices[i + 1])));
+            if (try pool.alloc(path_parents[next_d_offset..path_len])) {
```

Comment:
the current code allocates from `d_offset`, this allocates from `next_d_offset` because that's what we have in the `rebind()` api call below